automata/fa/dfa.py

```diff
@@ -540,44 +540,61 @@ def _stringify_states(states):
         return '{{{}}}'.format(','.join(sorted(str(state) for state in states)))
 
     @classmethod
-    def from_nfa(cls, target_nfa):
+    def from_nfa(cls, target_nfa, retain_names=False):
         """Initialize this DFA as one equivalent to the given NFA."""
         dfa_states = set()
         dfa_symbols = target_nfa.input_symbols
-        dfa_transitions = dict()
+        dfa_transitions = {}
+
+        # Data structures for state renaming
+        new_state_name_dict = dict()
+        state_name_counter = count(0)
+
+        def get_name_renamed(states):
+            nonlocal state_name_counter, new_state_name_dict
+            return new_state_name_dict.setdefault(frozenset(states), next(state_name_counter))
```

Comment:
@eliotwrobson This looks a lot like what I was _trying_ to do with `NFA._add_new_state`, but just much better executed lol.

Is there a reason you chose to use `itertools.count` as the counter variable, rather than just an `int` value and `+=` incrementing logic? I've never used `itertools.count` before, so I'm mostly just curious what use cases it was created to solve.

Reply:
Itertools count is just for situations where you want to be able to increment an integer counter and return the value of that counter each time (i.e. exactly this situation). This function originally used an int value with incrementing logic, but using `count` and calling the `next` function is basically just to get a one-line solution (which I realized you could do as I was copying it over).

Reply:
@eliotwrobson I mean, it's certainly cleaner, and seemingly more declarative too (e.g. no fussing about whether the `+=` line happens _before_ or _after_ where the retrieval occurs). I like it!